dockertest/environment.py

```diff
@@ -79,3 +81,95 @@ def selinux_is_enforcing():
         raise ValueError("Unexpected value from"
                          " security_getenforce(): %s" % mode)
     return mode == selinux.ENFORCING
+
+
+class RPMS(Mapping):
+    """
+    Read-only mapping of rpm names to ``nvra_type`` instances
+
+    :param args: Same as standard ``dict()`` positional arguments.
+    :param dargs: Same as standard ``dict()`` keyword arguments.
+    """
+
+    #: Class to use for representing nvra values, the first index
+    #: must be the package name.
+    nvra_type = namedtuple('nvra_type',
+                           ('name', 'version', 'release', 'arch'))
+    nvra_len = len(nvra_type._fields)
+
+    #: Query command to use for looking up a package NVRA must
+    #: return line-delimited items of hash-delimited fields
+    #: suitable for initializing ``nvra_type``.
+    rpmcmd = ("/usr/bin/rpm", "-q", "--qf", '%{N}#%{V}#%{R}#%{ARCH}\n')
+
+    #: Callable to use for executing rpmcmd
+    execute = staticmethod(subprocess.check_output)
+
+    #: Private, do not use
+    _cache = None
+    _alldone = None
+
+    def __new__(cls, *args, **dargs):
+        del args, dargs  # not used
+        if cls._cache is None:
+            cls.flush()
+        return object.__new__(cls)
+
+    def __init__(self, *args, **dargs):
+        # Support dict API
+        if not self._cache and (args or dargs):
+            argtest = dict(*args, **dargs)
+            for key, value in argtest.iteritems():
+                try:
+                    self._cache[key] = self.nvra_type(*value)
+                except TypeError:
+                    raise TypeError("Initializer key %s's value %s"
+                                    " incompatible with %s"
+                                    % (key, value, self.nvra_type))
+
+    @classmethod
+    def _store(cls, name):
+        cmd = list(cls.rpmcmd) + [name]
+        xcept_fmt = ("%s did not return expected line-delimited"
+                     " items of hash-delimited fields as"
+                     " output: %s")
+        output = cls.execute(cmd)
+        for value in output.strip().splitlines():
+            # Use a field-delimiter incompatible with any field value
+            new_nvra = cls.nvra_type(*value.strip().split('#'))
+            if len(new_nvra) != cls.nvra_len:
```

Comment:
It can, if the storage class (``nvra_type``) doesn't behave like the default ``namedtuple``.  This is done in nearly every unittest, where ``nvra_type`` is overloaded to be a regular tuple.  Probably a more complete way to validate the instantiation parameters,  would be to have the checker itself be mutable.  In anycase, I think that can be a detail left up to someone else to fix if ever needed.